### PR TITLE
Add permission for Accelerometer, Gyroscope, Magnetometer and Orientation Sensor

### LINKED ISSUE #141: Add permission for Accelerometer, Gyroscope, Magnetometer and Orientation Sensor.
There was consensus to support granular permissions in https://github.com/w3c/sensors/issues/22#issuecomment-285663269

Related PRs in the sensor specs are as follows : 
https://github.com/w3c/orientation-sensor/pull/27
https://github.com/w3c/magnetometer/pull/23
https://github.com/w3c/gyroscope/pull/12
https://github.com/w3c/accelerometer/pull/16

### PR DESCRIPTION
**Pull request by riju**
Fixes #141.

### CODE REVIEW

**Comment by riju** [comment]
@tobie : PTAL.

**Comment by tobie** [comment]
I think this needs more discussion. What's the relationship between those permissions? i.e. what happens when the orientation-sensor permission is granted and then further along the gyroscope permission is requested? Etc.

**Comment by anssiko** [comment]
Currently no relationship, each one is considered separately. Based on the learnings from the recent [WebVR Origin Trial][1], I feel we actually should ship this as an experiment in order to have definitive answer to this kind of ergonomics issues.

I leave it to the editors to decide whether to land this spec patch now and possibly adjust later, or leave this open until we have gathered such web developer feedback by shipping an experiment.

[1]: https://groups.google.com/a/chromium.org/d/msg/blink-dev/c41q3tyCBJE/I3zWysiXCgAJ

**Comment by tobie** [comment]
> I feel we actually should ship this as an experiment in order to have definitive answer to this kind of ergonomics issues.

I doubt you'll have anything clear as an answer here unless you ship *all* motion sensors API together.

**Comment by riju** [comment]
Something concrete as proposed by the Chrome Privacy/Security team after some analysis is as follows:  

AMBIENT_LIGHT_SENSOR: opt-out UI. granted by default.
MAGNETOMETER: opt-out UI. granted by default.
ORIENTATION_SENSOR: opt-out UI. granted by default.

ACCELEROMETER: auto-grant. no UI.
GYROSCOPE: auto grant. no UI.

We would go into origin trial with this model, get feedback from developers, and are open to change the policy based on feedback.

**Comment by riju** [comment]
Regarding this PR, iirc it was agreed in TPAC that we use granular permission names.
Details : https://github.com/w3c/sensors/issues/22#issuecomment-285663269

So that implementation can later use whatever privacy policy them deem fit on top of this
https://github.com/w3c/sensors/issues/174#issuecomment-285409411

**Comment by tobie** [comment]
> Regarding this PR, iirc it was agreed in TPAC that we use granular permission names.

Yes. That was the conclusion then. There's now evidence that this probably won't work well in practice, especially for motion sensors, will cause issues with new APIs (such as revoking permissions), and needs further research/work.

**Comment by anssiko** [comment]
>I doubt you'll have anything clear as an answer here unless you ship all motion sensors API together. 

That's the plan for the Chrome Origin Trial.

@marcoscaceres any concerns? We'd like to land this to have it documented before we enter Origin Trial to gather web developers' feedback on the model. Follows https://github.com/w3c/permissions/commit/4f95684c0da8aa4feeef4ce66262b778515dfd12

**Comment by tobie** [comment]
>> I doubt you'll have anything clear as an answer here unless you ship all motion sensors API together.

>That's the plan for the Chrome Origin Trial.

I meant *all* motion sensors not just these fours.

**Comment by marcoscaceres** [comment]
@anssiko, maybe we can add a note to the spec to say that these are under an origin trial - and that we are testing how the combination works together? @tobie, would that help a bit?

**Comment by anssiko** [comment]
@marcoscaceres, that is indeed a very good idea.

@tobie, please propose a good wording for the note, perhaps linking to your related research and/or the issue in w3c/sensors so people know where to go to provide further feedback on the topic.

**Comment by anssiko** [comment]
@riju, please add the following note below the [enum PermissionName block][1]:

```
Note: The enumeration values {{accelerometer}}, {{gyroscope}}, {{magnetometer}}, and
{{orientation-sensor}} are considered provisional and are subject to change based on
feedback from early implementations. See
<a href="https://github.com/w3c/sensors/issues/22">w3c/sensors#22</a> issue for
more information.
```

With this addition we should be good to land this PR, right @tobie and @marcoscaceres?

[1]: https://w3c.github.io/permissions/#permission-registry

**Comment by tobie** [comment]
LGTM.

File: index.bs [hunk 1]
```diff
@@ -988,6 +999,62 @@ spec: webidl
...
+    </p>
+  </section>
+  <section>
+    <h3 id="magnetometer">
+      Magnetometer
+    </h3>
+    <p>
+      The <dfn for="PermissionName" enum-value>"magnetometer"</dfn>
+      permission is the permission associated with the usage of the
+      [[ambient-light]] API.
```
Comment (raymeskhoury) [inline index.bs]: I think this should be magnetometer?

File: index.bs [hunk 2]
```diff
@@ -988,6 +999,62 @@ spec: webidl
...
+    </p>
+  </section>
+  <section>
+    <h3 id="orientation-sensor">
+      Orientation Sensor
+    </h3>
+    <p>
+      The <dfn for="PermissionName" enum-value>"orientation-sensor"</dfn>
+      permission is the permission associated with the usage of the
+      [[ambient-light]] API.
```
Comment (raymeskhoury) [inline index.bs]: orientation sensor?

**Comment by pozdnyakov** [comment]
@riju think it's worth dropping "orientation-sensor" from the list due to https://github.com/w3c/orientation-sensor/issues/33

File: index.bs [hunk 3]
```diff
@@ -669,9 +669,19 @@ spec: webidl
...
       <a enum-value>"ambient-light-sensor"</a>,
+      <a enum-value>"accelerometer"</a>,
+      <a enum-value>"gyroscope"</a>,
+      <a enum-value>"magnetometer"</a>,
       <a enum-value>"clipboard"</a>,
     };
   </pre>
+  <p class="note">
+  The enumeration values {{accelerometer}}, {{gyroscope}}, {{magnetometer}}, and
+  {{orientation-sensor}} are considered provisional and are subject to change
```
Comment (raymeskhoury) [inline index.bs]: nit: this still mentions orientation-sensor

**Comment by anssiko** [comment]
@marcoscaceres, thanks for your patience with this one.

We now have LGTMs from @tobie and me, and all the review comments from @raymeskhoury have been addressed. @riju squashed the multiple commits into one so you can maintain your beautiful commit history :-)

@marcoscaceres, feel free to land this so we can advance with the implementation.

**Comment by anssiko** [comment]
Thanks @marcoscaceres! We owe you a 🍻

**Comment by marcoscaceres** [comment]
I'll be taking you up on that at TPAC! 💃

**Pull request merged**